### PR TITLE
Clarify package installer interface with abstract base class, and Pip/Conda subclasses

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This makes an Abstract interface that the package installers in the qt plugin dialog need to implement, then refactors out all the if/then conditionals of the old Installer class into two pip/conda subclasses.  Also adds (new) tests for both cases.

work-in-progress because I need to update the cancel button and the output widget.

cc @jaimergp @goanpeca 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
Awesome @tlambert03. Love this :) 

Should I start a review now or wait a bit it goes out of WIP ?

**Comment by tlambert03** [comment]
i think it's ready for review (just removed the WIP), thanks!  I fixed the bits I knew I needed to fix (the cancel button, and the stdout logging).  

One major difference you'll note here is that we now only have a single QProcess, with a queue... rather than multiple parallel QProcesses.  I know that will be slower in some cases, but I think it's probably safer not to have concurrent installation processes (which may have conflicting dependencies?).

looks like the windows test fail is real, but it's a new test, so might have always "been there"?

**Comment by jaimergp** [comment]
Conda doesn't do well under parallel jobs. There are issues with the lockfiles on the cache directories, for example. So a sequential queue sounds good.It would be even better if it could batch installs together.

**Comment by tlambert03** [comment]
> It would be even better if it could batch installs together.

yeah, this would be very possible: for instance, if a job is currently running, but then you press the "install" button on two different other plugins, it could batch them together for the next job.  that would make the "cancel" logic slightly more complex (as it would have to "unbatch"), but I think it's the right way to go.

**edit**: maybe on second thought: it would be better to have a bunch of checkboxes, rather than buttons, and one button that says "install checked".  and then every checked one gets installed in a batch, but you only have the option to cancel them all together?  (that would also make it more obvious when one failing plugin in a batch causes the whole batch to fail)

**Comment by jaimergp** [comment]
Yes, from the UI perspective a multi-selection mode via checkboxes makes most sense. I just want to make sure the queue can provide more than one package at a time. In other words, the queue should return an iterable of package names/specs, instead of a single item. Does that make sense?

**Comment by tlambert03** [comment]
Yep, that's already what the queue does.  Each job has an iterable of specs.  It's mostly a _gui_ limitation that makes it hard to actually create a job with multiple specs (but that's what the installer api is expecting).  Checkboxes should fix that

File: napari/_qt/dialogs/_tests/test_installer_process.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    cmd = [str(tmp_path), '--no-setuptools', '--no-wheel', '--activators', '']
+    return virtualenv.cli_run(cmd)
+
+
+@pytest.fixture
+def tmp_conda_env(tmp_path):
+    import subprocess
+
+    subprocess.check_call(
```
Comment (Carreau) [inline napari/_qt/dialogs/_tests/test_installer_process.py:28]: This fails on windows saying that `conda` is not found (i think).

File: napari/_qt/dialogs/_tests/test_installer_process.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        job_id = installer.uninstall(['pip-install-test'])
+
+    for pth in tmp_virtualenv.creator.libs:
+        if (pth / 'pip_install_test').exists():
+            raise AssertionError('pip_install_test still installed')
+
+    assert not installer.hasJobs()
+
+
+def test_conda_installer(qtbot, tmp_conda_env: Path):
```
Comment (Carreau) [inline napari/_qt/dialogs/_tests/test_installer_process.py:74]: This test  seem to take 1min30 by itself:

```
============================= slowest 10 durations =============================
  73.42s call     napari/_qt/dialogs/_tests/test_installer_process.py::test_conda_installer
  16.49s setup    napari/_qt/dialogs/_tests/test_installer_process.py::test_conda_installer
```

Should we skip it on all but the latest python to limit CI usage ?

Comment (tlambert03) [reply napari/_qt/dialogs/_tests/test_installer_process.py:74]: Sure, but we definitely do need to be testing it somewhere

**Comment by jaimergp** [comment]
Hi @tlambert03 - are you planning on working on this further? I can happily take over if you want!

**Comment by tlambert03** [comment]
Nope, feel free to take and run with it. Thanks

**Comment by jaimergp** [comment]
Continued in #5124. Closing for clarity! Thanks!

**Pull request closed without merging**